### PR TITLE
Support for unicode characters in JsonConvert deserialization

### PR DESCRIPTION
**Pull request by jazzman55**
## Description
- Created GetNextChar() method in JsonConvert. It reads next byte(s) from jsonBytes and decodes into char, according to [UTF-8 encoding rules](https://en.wikipedia.org/wiki/UTF-8#Encoding) and advances jsonPos accordingly to the number of bytes read.
- Replaced simple cast byte to char with call to GetNextChar(), where needed.
- Created unit test DeserializeObjectWithStringContainingNonAsciiChars

## Motivation and Context
- Fixes nanoFramework/Home#1172

## How Has This Been Tested?
- Unit test: DeserializeObjectWithStringContainingNonAsciiChars, which creates an object with string property, sets some non-ascii characters to the string, then serializes and deserializes the object.
- Ran all other unit tests.
- Benchmark: nanoFramework.Json.Benchmark

## Types of changes
- [x] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [ ] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/master/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Json?pullRequest=268) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Json?pullRequest=268) before we can accept your contribution.<br/><br/>:x: jazzman55 [sign now](https://cla.dotnetfoundation.org/nanoframework/nanoFramework.Json?pullRequest=268)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/nanoframework/nanoFramework.Json?pullRequest=268) it.</sub>

**Review by josesimoes — COMMENTED** [review]
Nice! Any chance you can add a benchmark test with this? I'm curious on the impact.

**Comment by josesimoes** [comment]
@jazzman55 please merge from upstream/main to pass the failing check.

**Comment by josesimoes** [comment]
@jazzman55 please address Sonarcloud code smells too

File: nanoFramework.Json/JsonConvert.cs [hunk 1]
```diff
@@ -1036,6 +1036,22 @@ private static LexToken GetNextToken()
             return result;
         }
 
+        private static char GetNextChar()
+        {
+            //https://en.wikipedia.org/wiki/UTF-8#Encoding
+
+            if ((jsonBytes[jsonPos] & 0x80) == 0)
```
[the remark this answers is not in the document]
Comment (jazzman55) [reply nanoFramework.Json/JsonConvert.cs:1036]: Below are the benchmarks. GetNextChar is the current implementation. I also benchmarked a variant, where the method implementation is done inline (inline.txt) - it is a bit faster (especially in ComplexObjectAzureTwinPayload) but also ugly and breaks DRY rule:

                ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
                    : (jsonBytes[jsonPos] & 0x20) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 2) - 2, 2)[0]
                    : (jsonBytes[jsonPos] & 0x10) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 3) - 3, 3)[0]
                    : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];

[baseline.txt](https://github.com/nanoframework/nanoFramework.Json/files/9868959/baseline.txt)
[GetNextChar.txt](https://github.com/nanoframework/nanoFramework.Json/files/9868961/GetNextChar.txt)
[inline.txt](https://github.com/nanoframework/nanoFramework.Json/files/9868962/inline.txt)

**Comment by josesimoes** [comment]
Benchmark for inline

```text
Console export: ReferenceTypesDeserializationBenchmark benchmark class.
| MethodName                    | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------------------------- |
| IntArray                      | 5               | 600 ms  | 590 ms   | 610 ms   |
| ArrayList                     | 5               | 1076 ms | 1,070 ms | 1,080 ms |
| ComplexObjectAzureTwinPayload | 5               | 544 ms  | 540 ms   | 550 ms   |
| ShortArray                    | 5               | 478 ms  | 470 ms   | 490 ms   |
| String                        | 5               | 150 ms  | 140 ms   | 160 ms   |
| NestedClass                   | 5               | 216 ms  | 210 ms   | 220 ms   |
| ComplexArrayObject            | 5               | 816 ms  | 810 ms   | 820 ms   |


Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean   | Min    | Max    |
| ------------------------------------------------------- |
| Short      | 5               | 168 ms | 160 ms | 180 ms |
| TimeSpanT  | 5               | 276 ms | 270 ms | 280 ms |
| Float      | 5               | 180 ms | 180 ms | 180 ms |
| Double     | 5               | 196 ms | 190 ms | 200 ms |
| DateTimeT  | 5               | 244 ms | 240 ms | 260 ms |
| Long       | 5               | 198 ms | 190 ms | 200 ms |
| EnumBoxed  | 5               | 314 ms | 310 ms | 320 ms |


Console export: Person benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassChild benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassComplex benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestCompany benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestEmployee benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestTown benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName         | ItterationCount | Mean   | Min    | Max    |
| --------------------------------------------------------------- |
| IntArray           | 5               | 144 ms | 140 ms | 150 ms |
| ShortArray         | 5               | 136 ms | 130 ms | 140 ms |
| String             | 5               | 54 ms  | 50 ms  | 60 ms  |
| NestedClass        | 5               | 158 ms | 150 ms | 160 ms |
| ComplexObject      | 5               | 256 ms | 250 ms | 260 ms |
| ComplexArrayObject | 5               | 454 ms | 450 ms | 460 ms |
| ArrayList          | 5               | 414 ms | 400 ms | 420 ms |


Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------ |
| Short      | 5               | 140 ms  | 130 ms   | 150 ms   |
| TimeSpanT  | 5               | 130 ms  | 130 ms   | 130 ms   |
| Float      | 5               | 184 ms  | 180 ms   | 190 ms   |
| Double     | 5               | 160 ms  | 160 ms   | 160 ms   |
| DateTimeT  | 5               | 1346 ms | 1,330 ms | 1,360 ms |
| Long       | 5               | 148 ms  | 140 ms   | 150 ms   |
| Enum       | 5               | 164 ms  | 160 ms   | 170 ms   |
| EnumBoxed  | 5               | 628 ms  | 620 ms   | 630 ms   |


Console export: TwinPayload benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |
```

**Comment by josesimoes** [comment]
Benchmark for GetNextChar (proposed implementation in this PR)

```text
Console export: ReferenceTypesDeserializationBenchmark benchmark class.
| MethodName                    | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------------------------- |
| IntArray                      | 5               | 614 ms  | 600 ms   | 620 ms   |
| ArrayList                     | 5               | 1166 ms | 1,160 ms | 1,170 ms |
| ComplexObjectAzureTwinPayload | 5               | 658 ms  | 650 ms   | 660 ms   |
| ShortArray                    | 5               | 494 ms  | 490 ms   | 500 ms   |
| String                        | 5               | 150 ms  | 150 ms   | 150 ms   |
| NestedClass                   | 5               | 238 ms  | 230 ms   | 240 ms   |
| ComplexArrayObject            | 5               | 890 ms  | 880 ms   | 900 ms   |


Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean   | Min    | Max    |
| ------------------------------------------------------- |
| Short      | 5               | 172 ms | 170 ms | 180 ms |
| TimeSpanT  | 5               | 300 ms | 290 ms | 310 ms |
| Float      | 5               | 188 ms | 180 ms | 190 ms |
| Double     | 5               | 200 ms | 200 ms | 200 ms |
| DateTimeT  | 5               | 274 ms | 270 ms | 280 ms |
| Long       | 5               | 200 ms | 200 ms | 200 ms |
| EnumBoxed  | 5               | 318 ms | 310 ms | 330 ms |


Console export: Person benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassChild benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassComplex benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestCompany benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestEmployee benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestTown benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName         | ItterationCount | Mean   | Min    | Max    |
| --------------------------------------------------------------- |
| IntArray           | 5               | 140 ms | 130 ms | 150 ms |
| ShortArray         | 5               | 136 ms | 130 ms | 140 ms |
| String             | 5               | 56 ms  | 50 ms  | 60 ms  |
| NestedClass        | 5               | 160 ms | 160 ms | 160 ms |
| ComplexObject      | 5               | 252 ms | 250 ms | 260 ms |
| ComplexArrayObject | 5               | 456 ms | 450 ms | 460 ms |
| ArrayList          | 5               | 400 ms | 400 ms | 400 ms |


Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------ |
| Short      | 5               | 130 ms  | 120 ms   | 140 ms   |
| TimeSpanT  | 5               | 128 ms  | 120 ms   | 130 ms   |
| Float      | 5               | 184 ms  | 180 ms   | 190 ms   |
| Double     | 5               | 160 ms  | 160 ms   | 160 ms   |
| DateTimeT  | 5               | 1328 ms | 1,320 ms | 1,350 ms |
| Long       | 5               | 142 ms  | 140 ms   | 150 ms   |
| Enum       | 5               | 158 ms  | 150 ms   | 160 ms   |
| EnumBoxed  | 5               | 600 ms  | 590 ms   | 610 ms   |


Console export: TwinPayload benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |

```

**Comment by josesimoes** [comment]
Baseline benchmark

```text
Console export: ReferenceTypesDeserializationBenchmark benchmark class.
| MethodName                    | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------------------------- |
| IntArray                      | 5               | 590 ms  | 580 ms   | 600 ms   |
| ArrayList                     | 5               | 1046 ms | 1,030 ms | 1,060 ms |
| ComplexObjectAzureTwinPayload | 5               | 520 ms  | 520 ms   | 520 ms   |
| ShortArray                    | 5               | 468 ms  | 460 ms   | 480 ms   |
| String                        | 5               | 144 ms  | 140 ms   | 150 ms   |
| NestedClass                   | 5               | 208 ms  | 200 ms   | 210 ms   |
| ComplexArrayObject            | 5               | 792 ms  | 790 ms   | 800 ms   |


Console export: ValueTypesDeserializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean   | Min    | Max    |
| ------------------------------------------------------- |
| Short      | 5               | 166 ms | 160 ms | 170 ms |
| TimeSpanT  | 5               | 262 ms | 260 ms | 270 ms |
| Float      | 5               | 176 ms | 170 ms | 180 ms |
| Double     | 5               | 190 ms | 180 ms | 200 ms |
| DateTimeT  | 5               | 226 ms | 220 ms | 230 ms |
| Long       | 5               | 186 ms | 180 ms | 190 ms |
| EnumBoxed  | 5               | 300 ms | 290 ms | 310 ms |


Console export: Person benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassChild benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestClassComplex benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestCompany benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestEmployee benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: JsonTestTown benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |


Console export: ReferenceTypesSerializationBenchmark benchmark class.
| MethodName         | ItterationCount | Mean   | Min    | Max    |
| --------------------------------------------------------------- |
| IntArray           | 5               | 134 ms | 130 ms | 140 ms |
| ShortArray         | 5               | 134 ms | 130 ms | 140 ms |
| String             | 5               | 56 ms  | 50 ms  | 60 ms  |
| NestedClass        | 5               | 156 ms | 150 ms | 160 ms |
| ComplexObject      | 5               | 250 ms | 240 ms | 260 ms |
| ComplexArrayObject | 5               | 444 ms | 430 ms | 450 ms |
| ArrayList          | 5               | 402 ms | 390 ms | 410 ms |


Console export: ValueTypesSerializationBenchmark benchmark class.
| MethodName | ItterationCount | Mean    | Min      | Max      |
| ------------------------------------------------------------ |
| Short      | 5               | 130 ms  | 120 ms   | 140 ms   |
| TimeSpanT  | 5               | 122 ms  | 120 ms   | 130 ms   |
| Float      | 5               | 180 ms  | 180 ms   | 180 ms   |
| Double     | 5               | 152 ms  | 150 ms   | 160 ms   |
| DateTimeT  | 5               | 1326 ms | 1,320 ms | 1,330 ms |
| Long       | 5               | 142 ms  | 130 ms   | 160 ms   |
| Enum       | 5               | 156 ms  | 150 ms   | 160 ms   |
| EnumBoxed  | 5               | 604 ms  | 600 ms   | 620 ms   |


Console export: TwinPayload benchmark class.
| MethodName | ItterationCount | Mean | Min | Max |
| ----------------------------------------------- |

```

File: nanoFramework.Json/JsonConvert.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (josesimoes) [reply nanoFramework.Json/JsonConvert.cs:1036]: We don't have to be fundamentalists here. 😅 
I'm quite all right with breaking the DRY here and have that "ugly" code. Performance it's paramount in this case!

I'm good with this changes and I would go for the inline version, considering it's faster.

@torbacz @Ellerbach any comments on this?

**Comment by torbacz** [comment]
All results in one table
image

**Review by torbacz — CHANGES_REQUESTED** [review]

**Review by torbacz — APPROVED** [review]

**Comment by torbacz** [comment]
@josesimoes 
Any idea how can we fix sonar?

**Comment by jazzman55** [comment]
I found a possible way of inlining a method[inlining a method](https://www.dotnetperls.com/aggressiveinlining), which theoretically would get us the best of both worlds - performance and keep sonar happy. Does nano CLR support this attribute [MethodImpl(MethodImplOptions.AggressiveInlining] ?

**Comment by josesimoes** [comment]
> I found a possible way of inlining a method[inlining a method](https://www.dotnetperls.com/aggressiveinlining), which theoretically would get us the best of both worlds - performance and keep sonar happy. Does nano CLR support this attribute [MethodImpl(MethodImplOptions.AggressiveInlining] ?

I'm afraid we don't have that attribute available.
From a quick search it seems that this is handled by the compiler, so it could be a matter of adding support for it.
I can investigate this tomorrow.

If that fails, we can always ignore Sonarcloud analysis for this code block.

File: nanoFramework.Json/JsonConvert.cs [hunk 2]
```diff
@@ -1043,7 +1042,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
                     return EndToken(sb);
                 }
 
-                ch = (char)jsonBytes[jsonPos++];
+                ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1045]:
```suggestion
#pragma warning disable S3881
                ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```

File: nanoFramework.Json/JsonConvert.cs [hunk 3]
```diff
@@ -1043,7 +1042,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
...
+                    : (jsonBytes[jsonPos] & 0x20) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 2) - 2, 2)[0]
+                    : (jsonBytes[jsonPos] & 0x10) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 3) - 3, 3)[0]
+                    : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1048]:
```suggestion
                    : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
#pragma warning restore S3358
```

File: nanoFramework.Json/JsonConvert.cs [hunk 4]
```diff
@@ -1052,7 +1054,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
                 if (ch == '\\')
                 {
                     escaped = true;
-                    ch = (char)jsonBytes[jsonPos++];
+                    ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1057]:
```suggestion
#pragma warning disable S3881
                    ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```

File: nanoFramework.Json/JsonConvert.cs [hunk 5]
```diff
@@ -1052,7 +1054,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
...
+                        : (jsonBytes[jsonPos] & 0x20) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 2) - 2, 2)[0]
+                        : (jsonBytes[jsonPos] & 0x10) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 3) - 3, 3)[0]
+                        : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1060]:
```suggestion
                        : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
#pragma warning restore S3881
```

File: nanoFramework.Json/JsonConvert.cs [hunk 6]
```diff
@@ -1262,7 +1267,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
 
         private static void Expect(char expected, ref int jsonPos, ref byte[] jsonBytes)
         {
-            char ch = (char)jsonBytes[jsonPos++];
+            char ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
+                : (jsonBytes[jsonPos] & 0x20) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 2) - 2, 2)[0]
+                : (jsonBytes[jsonPos] & 0x10) == 0 ? Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 3) - 3, 3)[0]
+                : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1273]:
```suggestion
                : Encoding.UTF8.GetChars(jsonBytes, (jsonPos += 4) - 4, 4)[0];
#pragma warning restore S3358
```

File: nanoFramework.Json/JsonConvert.cs [hunk 7]
```diff
@@ -1262,7 +1267,10 @@ private static LexToken GetNextTokenInternal(ref int jsonPos, ref byte[] jsonByt
 
         private static void Expect(char expected, ref int jsonPos, ref byte[] jsonBytes)
         {
-            char ch = (char)jsonBytes[jsonPos++];
+            char ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```
Comment (torbacz) [inline nanoFramework.Json/JsonConvert.cs:1270]:
```suggestion
#pragma warning disable S3881
            char ch = (jsonBytes[jsonPos] & 0x80) == 0 ? (char)jsonBytes[jsonPos++]
```

**Review by torbacz — COMMENTED** [review]
Should suppress sonar issues.

**Comment by josesimoes** [comment]
Yes. that should take care of it. @jazzman55 care to merge the suggestions please?

**Review by josesimoes — APPROVED** [review]
LGTM! 👍🏻 💯

**Pull request merged**